# Who will win the 2020 Masters?

Analysis is paused til next quarter.  Will revisit then!


## Data set
I chose PGA Tour ShotLink data from 1980 to 2019 (www.shotlink.com and www.pgatour.com).  This is the largest set of data available on these metrics.  The data is manually collected by volunteers who follow every PGA Tour player and record every shot made in every PGA Tour tournament.  I was looking into scraping the data when I realized someone on Kaggle had already done this (https://www.kaggle.com/bradklassen/pga-tour-20102018-data).  I instead used this data set, formatted it, and cleaned out obvious errors.


## Conclusion
TBD!


# Detailed steps shown below

## Importing, Formatting and Cleaning the Data

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
from scipy import stats
sns.set_palette("GnBu_d")

In [2]:
df = pd.read_csv('pga_data_historical.csv') #2010-2019 PGA Tour data, source: https://www.kaggle.com/bradklassen/pga-tour-20102018-data (as of August 2020, he has taken the data down)

In [3]:
df.head()

player_name        date               tournament  \
0  Rik Massengale  1980-01-13  Bob Hope Desert Classic   
1   Bobby Nichols  1980-01-13  Bob Hope Desert Classic   
2      Andy North  1980-01-13  Bob Hope Desert Classic   
3   John Mahaffey  1980-01-13  Bob Hope Desert Classic   
4  Peter Jacobsen  1980-01-13  Bob Hope Desert Classic   

                     statistic variable  value  
0  Final Round Scoring Average      AVG  70.00  
1  Final Round Scoring Average      AVG  73.00  
2  Final Round Scoring Average      AVG  73.00  
3  Final Round Scoring Average      AVG  73.00  
4  Final Round Scoring Average      AVG  73.00

In [4]:
df.shape

(46147897, 6)

In [5]:
#Unearthing the available variables in the "statistic" column
df.statistic.unique()

array(['Final Round Scoring Average', 'All-Around Ranking',
       'Par 4 Birdie or Better Leaders', 'Scoring Average Before Cut',
       'Scoring Average (Actual)', 'Round 3 Scoring Average',
       'Official Money', 'Greens in Regulation Percentage',
       'Total Birdies', 'Birdie or Better Conversion Percentage',
       'Eagles (Holes per)', 'Total Driving', 'Scoring Average',
       'Par 3 Birdie or Better Leaders', 'Driving Distance',
       'Sand Save Percentage', 'Putts Per Round',
       'Percentage of potential money won', 'Driving Accuracy Percentage',
       'Par 5 Birdie or Better Leaders', 'Par Breakers', 'Ball Striking',
       'Birdie Average', 'Total Money (Official and Unofficial)',
       'Total Eagles', 'Front 9 Round 1 Scoring Average',
       'Back 9 Round 1 Scoring Average', 'Back 9 Par 5 Scoring Average',
       'Front 9 Round 2 Scoring Average', 'Front 9 Par 4 Scoring Average',
       'Back 9 Par 3 Scoring Average', 'Front 9 Par 3 Scoring Average',
       'Fron

In [6]:
df.statistic.value_counts()

Total Eagles                               256346
Total Money (Official and Unofficial)      256156
Total Birdies                              255870
Percentage of potential money won          254714
Final Round Scoring Average                252576
                                            ...  
Late Round 5 Scoring Average                  890
Tenth Tee Early Round 5 Scoring Average       528
First Tee Late Round 5 Scoring Average        468
Tenth Tee Late Round 5 Scoring Average        422
First Tee Early Round 5 Scoring Average       276
Name: statistic, Length: 442, dtype: int64

In [7]:
#Selecting from the hundreds of variables the "statistic" column

chosen_statistic = ['Greens in Regulation Percentage',
                    'Driving Accuracy Percentage',
                    'Scrambling',
                    'Scoring Average (Actual)',
                    'Official Money',
                    'Percentage of Available Purse Won',
                    'Driving Distance',
                    'Sand Save Percentage',
                    'Overall Putting Average']

#Definitions for each variable:
    #Greens in Regulation Percentage: The percent of time a player was able to hit the green in regulation (greens hit in regulation/holes played). Note: A green is considered hit in regulation if any portion of the ball is touching the putting surface after the GIR stroke has been taken. (The GIR stroke is determined by subtracting 2 from par (1st stroke on a par 3, 2nd on a par 4, 3rd on a par 5))
    #Driving Accuracy Percentage: The percentage of time a tee shot comes to rest in the fairway (regardless of club)
    #Scrambling: The percent of time a player misses the green in regulation, but still makes par or better.
    #Scoring Average (Actual): The average number of strokes per completed round.
    #Official Money: The total official money a player has earned year-to-date. Note: This is for PGA TOUR members only.
    #Percentage of Available Purse Won: For official events, the player's total money won as a percentage of the total purse available.
    #Driving Distance: The average number of yards per measured drive. These drives are measured on two holes per round. Care is taken to select two holes which face in opposite directions to counteract the effect of wind. Drives are measured to the point at which they come to rest regardless of whether they are in the fairway or not.
    #Sand Save Percentage: The percent of time a player was able to get 'up and down' once in a greenside sand bunker (regardless of score). Note: 'Up and down' indicates it took the player 2 shots or less to put the ball in the hole from that point.
    #Overall Putting Average: The average number of putts for all holes played (total putts / total holes played).

In [8]:
#Adjusting the DataFrame
df = df[df['statistic'].isin(chosen_statistic)]
df = df[df['variable'] != 'RANK THIS WEEK'] #Multiple variables have the value 'RANK THIS WEEK', so unstacking will create duplicate columns of 'RANK THIS WEEK', causing an error ("Index contains duplicate entries, cannot reshape"). This line eliminates the error
df = df.drop(columns = ['variable'])
df.head()

player_name        date               tournament  \
145     Bill Rogers  1980-01-13  Bob Hope Desert Classic   
146      Gil Morgan  1980-01-13  Bob Hope Desert Classic   
147      Bob Gilder  1980-01-13  Bob Hope Desert Classic   
148  Billy Kratzert  1980-01-13  Bob Hope Desert Classic   
149   Roger Maltbie  1980-01-13  Bob Hope Desert Classic   

                    statistic  value  
145  Scoring Average (Actual)  70.60  
146  Scoring Average (Actual)  70.60  
147  Scoring Average (Actual)  70.60  
148  Scoring Average (Actual)  70.60  
149  Scoring Average (Actual)  70.60

In [9]:
#Unstacking the data in the DataFrame
df = df.set_index(['player_name', 'date', 'tournament','statistic'])['value'].unstack('statistic').reset_index()
df.head()

statistic     player_name        date                            tournament  \
0             A.J. Duncan  1988-04-03         KMart Greater Greensboro Open   
1             A.J. Duncan  1990-03-11                         Honda Classic   
2          A.J. McInerney  2017-11-05  Shriners Hospitals for Children Open   
3          A.J. McInerney  2018-06-10                FedEx St. Jude Classic   
4          Aaron Baddeley  2000-03-12                         Honda Classic   

statistic Driving Accuracy Percentage Driving Distance  \
0                               58.93            283.5   
1                               44.64            264.1   
2                               50.00            317.8   
3                               50.00            298.5   
4                               64.29            283.6   

statistic Greens in Regulation Percentage Official Money  \
0                                   58.33         $1,810   
1                                   52.78         $1,810   
2                                   76.39            NaN   
3                                   56.94            NaN   
4                                   70.83            NaN   

statistic Overall Putting Average Percentage of Available Purse Won  \
0                             NaN                               .18   
1                             NaN                               .18   
2                           1.694                              2.22   
3                           1.514                               .22   
4                           1.625                               NaN   

statistic Sand Save Percentage Scoring Average (Actual) Scrambling  
0                        33.33                    73.75        NaN  
1                        14.29                    77.00        NaN  
2                        22.22                    69.50      52.94  
3                        33.33                    70.75      61.29  
4                        50.00                    70.25      66.67

In [10]:
df.shape

(128160, 12)

In [11]:
df.describe()

statistic     player_name        date        tournament  \
count              128160      128160            128160   
unique               2440        1697               304   
top        Davis Love III  1995-07-23  PGA Championship   
freq                  546         178              3018   

statistic Driving Accuracy Percentage Driving Distance  \
count                          124669           123306   
unique                            333             1169   
top                             64.29            271.9   
freq                             7030              322   

statistic Greens in Regulation Percentage Official Money  \
count                              124543         123674   
unique                                111          11277   
top                                 66.67              $   
freq                                 8056            200   

statistic Overall Putting Average Percentage of Available Purse Won  \
count                       87568                            118148   
unique                        206                               772   
top                         1.611                               .22   
freq                         6397                              7025   

statistic Sand Save Percentage Scoring Average (Actual) Scrambling  
count                   123011                   126125      87574  
unique                     134                      129        425  
top                      50.00                    70.00      66.67  
freq                     19184                     6630       4253

## Fixing the Data types

In [12]:
print(df.dtypes)

statistic
player_name                          object
date                                 object
tournament                           object
Driving Accuracy Percentage          object
Driving Distance                     object
Greens in Regulation Percentage      object
Official Money                       object
Overall Putting Average              object
Percentage of Available Purse Won    object
Sand Save Percentage                 object
Scoring Average (Actual)             object
Scrambling                           object
dtype: object


In [13]:
df.head()

statistic     player_name        date                            tournament  \
0             A.J. Duncan  1988-04-03         KMart Greater Greensboro Open   
1             A.J. Duncan  1990-03-11                         Honda Classic   
2          A.J. McInerney  2017-11-05  Shriners Hospitals for Children Open   
3          A.J. McInerney  2018-06-10                FedEx St. Jude Classic   
4          Aaron Baddeley  2000-03-12                         Honda Classic   

statistic Driving Accuracy Percentage Driving Distance  \
0                               58.93            283.5   
1                               44.64            264.1   
2                               50.00            317.8   
3                               50.00            298.5   
4                               64.29            283.6   

statistic Greens in Regulation Percentage Official Money  \
0                                   58.33         $1,810   
1                                   52.78         $1,810   
2                                   76.39            NaN   
3                                   56.94            NaN   
4                                   70.83            NaN   

statistic Overall Putting Average Percentage of Available Purse Won  \
0                             NaN                               .18   
1                             NaN                               .18   
2                           1.694                              2.22   
3                           1.514                               .22   
4                           1.625                               NaN   

statistic Sand Save Percentage Scoring Average (Actual) Scrambling  
0                        33.33                    73.75        NaN  
1                        14.29                    77.00        NaN  
2                        22.22                    69.50      52.94  
3                        33.33                    70.75      61.29  
4                        50.00                    70.25      66.67

In [14]:
#Cleaning the "Official Money" column so we can analyze it
df['Official Money'] = df['Official Money'].str.replace('\$', '', regex = True)
df['Official Money'] = df['Official Money'].str.replace(',', '', regex = True)
df['Official Money'] = df['Official Money'].str.replace('O', '0', regex = True)
df['Official Money'] = df['Official Money'].str.replace(' ', '', regex = True)
df['Official Money'] = df['Official Money'].str.strip()
df['Official Money'].replace("", np.nan, inplace=True)

In [15]:
#Converting data types from objects to float
df['Greens in Regulation Percentage']=df['Greens in Regulation Percentage'].astype('float')
df['Driving Accuracy Percentage']=df['Driving Accuracy Percentage'].astype('float')
df['Scrambling']=df['Scrambling'].astype('float')
df['Scoring Average (Actual)']=df['Scoring Average (Actual)'].astype('float')
df['Official Money']=df['Official Money'].astype('float')
df['Percentage of Available Purse Won']=df['Percentage of Available Purse Won'].astype('float')
df['Driving Distance']=df['Driving Distance'].astype('float')
df['Sand Save Percentage']=df['Sand Save Percentage'].astype('float')
df['Overall Putting Average']=df['Overall Putting Average'].astype('float')

In [16]:
print(df.dtypes)

statistic
player_name                           object
date                                  object
tournament                            object
Driving Accuracy Percentage          float64
Driving Distance                     float64
Greens in Regulation Percentage      float64
Official Money                       float64
Overall Putting Average              float64
Percentage of Available Purse Won    float64
Sand Save Percentage                 float64
Scoring Average (Actual)             float64
Scrambling                           float64
dtype: object


In [17]:
#Cleaning up the DataFrame - renaming certain columns
df.rename(columns = {'Greens in Regulation Percentage':'GIR',
                     'Driving Accuracy Percentage':'Fairways',
                     'Scrambling':'Scrambling',
                     'Scoring Average (Actual)':'Score',
                     'Official Money':'Money',
                     'Percentage of Available Purse Won':'Purse',
                     'Driving Distance':'Distance',
                     'Sand Save Percentage':'Sandies',
                     'Overall Putting Average':'Putts'}, inplace=True)

In [18]:
df.head()

statistic     player_name        date                            tournament  \
0             A.J. Duncan  1988-04-03         KMart Greater Greensboro Open   
1             A.J. Duncan  1990-03-11                         Honda Classic   
2          A.J. McInerney  2017-11-05  Shriners Hospitals for Children Open   
3          A.J. McInerney  2018-06-10                FedEx St. Jude Classic   
4          Aaron Baddeley  2000-03-12                         Honda Classic   

statistic  Fairways  Distance    GIR   Money  Putts  Purse  Sandies  Score  \
0             58.93     283.5  58.33  1810.0    NaN   0.18    33.33  73.75   
1             44.64     264.1  52.78  1810.0    NaN   0.18    14.29  77.00   
2             50.00     317.8  76.39     NaN  1.694   2.22    22.22  69.50   
3             50.00     298.5  56.94     NaN  1.514   0.22    33.33  70.75   
4             64.29     283.6  70.83     NaN  1.625    NaN    50.00  70.25   

statistic  Scrambling  
0                 NaN  
1                 NaN  
2               52.94  
3               61.29  
4               66.67

In [19]:
df.describe()

statistic       Fairways       Distance            GIR         Money  \
count      124669.000000  123306.000000  124543.000000  1.234740e+05   
mean           65.607781     275.949496      66.555930  4.870611e+04   
std            10.898736      19.243917       8.459017  1.231114e+05   
min            17.860000     182.600000      20.830000  9.600000e+01   
25%            58.930000     261.500000      61.110000  4.265000e+03   
50%            66.070000     275.300000      66.670000  1.332500e+04   
75%            73.210000     289.800000      72.220000  3.987500e+04   
max           100.000000     362.100000     487.500000  2.250000e+06   

statistic         Putts          Purse        Sandies          Score  \
count      87568.000000  118148.000000  123011.000000  126125.000000   
mean           1.604443       1.407430      51.691787      70.651087   
std            0.084746       2.585814      24.156935       1.862798   
min            0.611000       0.130000       0.000000      63.250000   
25%            1.556000       0.230000      36.360000      69.250000   
50%            1.611000       0.520000      50.000000      70.500000   
75%            1.653000       1.450000      66.670000      71.750000   
max            2.056000      30.000000     200.000000      81.250000   

statistic    Scrambling  
count      87574.000000  
mean          61.097847  
std           11.346915  
min            4.170000  
25%           53.570000  
50%           61.290000  
75%           68.750000  
max          100.000000

### Fixing errors in the data
There is at least one GIR value that is above 100% (in the chart above the max = 487.5) and one Sandies value that is above 100% (in the chart above the max = 200.0).  Will drop any rows where GIR or Sandies is above 100%:

In [20]:
df=df[df.GIR <= 100]
df=df[df.Sandies <= 100]

In [21]:
df.describe()

statistic       Fairways       Distance            GIR         Money  \
count      122818.000000  121486.000000  122848.000000  1.184600e+05   
mean           65.584648     276.036268      66.498876  4.843005e+04   
std            10.904967      19.236936       8.347419  1.213998e+05   
min            17.860000     182.600000      20.830000  2.800000e+02   
25%            58.930000     261.500000      61.110000  4.332000e+03   
50%            66.070000     275.400000      66.670000  1.339200e+04   
75%            73.210000     289.900000      72.220000  3.990000e+04   
max           100.000000     362.100000     100.000000  2.250000e+06   

statistic         Putts          Purse        Sandies          Score  \
count      86385.000000  113663.000000  122848.000000  122691.000000   
mean           1.604132       1.391657      51.688042      70.651976   
std            0.084406       2.558169      24.150509       1.852941   
min            0.611000       0.130000       0.000000      63.250000   
25%            1.556000       0.230000      36.360000      69.330000   
50%            1.611000       0.520000      50.000000      70.500000   
75%            1.653000       1.450000      66.670000      71.750000   
max            2.056000      30.000000     100.000000      81.250000   

statistic    Scrambling  
count      86391.000000  
mean          61.093638  
std           11.312598  
min            4.170000  
25%           53.570000  
50%           61.290000  
75%           68.750000  
max          100.000000

In [22]:
#Adding the year as a column to the dataframe
df['year'] = pd.DatetimeIndex(df['date']).year
df['month'] = pd.DatetimeIndex(df['date']).month
df.head()

statistic     player_name        date                            tournament  \
0             A.J. Duncan  1988-04-03         KMart Greater Greensboro Open   
1             A.J. Duncan  1990-03-11                         Honda Classic   
2          A.J. McInerney  2017-11-05  Shriners Hospitals for Children Open   
3          A.J. McInerney  2018-06-10                FedEx St. Jude Classic   
4          Aaron Baddeley  2000-03-12                         Honda Classic   

statistic  Fairways  Distance    GIR   Money  Putts  Purse  Sandies  Score  \
0             58.93     283.5  58.33  1810.0    NaN   0.18    33.33  73.75   
1             44.64     264.1  52.78  1810.0    NaN   0.18    14.29  77.00   
2             50.00     317.8  76.39     NaN  1.694   2.22    22.22  69.50   
3             50.00     298.5  56.94     NaN  1.514   0.22    33.33  70.75   
4             64.29     283.6  70.83     NaN  1.625    NaN    50.00  70.25   

statistic  Scrambling  year  month  
0                 NaN  1988      4  
1                 NaN  1990      3  
2               52.94  2017     11  
3               61.29  2018      6  
4               66.67  2000      3

# Predicting the 2019 Masters winner

## Analyzing prior Masters Tournament winners

In [23]:
#How the 2019 winner performed in 2018
df_2019winner = df.loc[(df['player_name'] == 'Tiger Woods') & (df['year']==2018)]
df_2019winner.head(50)

statistic  player_name        date  \
117776     Tiger Woods  2018-01-28   
117777     Tiger Woods  2018-02-25   
117778     Tiger Woods  2018-03-11   
117779     Tiger Woods  2018-03-18   
117780     Tiger Woods  2018-03-18   
117781     Tiger Woods  2018-04-08   
117782     Tiger Woods  2018-05-06   
117783     Tiger Woods  2018-05-13   
117784     Tiger Woods  2018-06-03   
117785     Tiger Woods  2018-07-01   
117786     Tiger Woods  2018-07-22   
117787     Tiger Woods  2018-08-05   
117788     Tiger Woods  2018-08-12   
117789     Tiger Woods  2018-08-26   
117790     Tiger Woods  2018-09-03   
117791     Tiger Woods  2018-09-09   
117792     Tiger Woods  2018-09-23   

statistic                                         tournament  Fairways  \
117776                                Farmers Insurance Open     30.36   
117777                                     The Honda Classic     58.93   
117778                                  Valspar Championship     59.62   
117779     Arnold Palmer Invitational presented by Master...     60.71   
117780          Corales Puntacana Resort & Club Championship     60.71   
117781                                    Masters Tournament     53.57   
117782                              Wells Fargo Championship     44.64   
117783                              THE PLAYERS Championship     62.50   
117784       the Memorial Tournament presented by Nationwide     71.43   
117785                                Quicken Loans National     57.14   
117786                                 The Open Championship     66.67   
117787     World Golf Championships-Bridgestone Invitational     46.43   
117788                                      PGA Championship     57.14   
117789                                    THE NORTHERN TRUST     67.86   
117790                        Dell Technologies Championship     71.43   
117791                                      BMW Championship     76.79   
117792                                     TOUR Championship     64.29   

statistic  Distance    GIR      Money  Putts  Purse  Sandies  Score  \
117776        302.5  58.33    59685.0  1.528   0.87    40.00  71.25   
117777        318.9  66.67   151800.0  1.597   2.30    50.00  70.00   
117778        296.3  66.67   572000.0  1.542   8.80    42.86  68.75   
117779        306.8  62.50   338200.0  1.486   3.80    55.56  69.50   
117780        306.8  62.50   338200.0  1.486   3.80    55.56  69.50   
117781        293.8  66.67    63663.0  1.625   0.58    33.33  72.25   
117782        330.3  70.83    17479.0  1.750   0.23    14.29  71.50   
117783        297.3  69.44   225500.0  1.542   2.05    60.00  69.25   
117784        300.9  75.00    76985.0  1.639   0.87    75.00  69.75   
117785        305.9  73.61   312400.0  1.514   4.40    50.00  67.25   
117786        303.0  66.67   327000.0  1.639   3.11    50.00  69.75   
117787        302.5  62.50    74750.0  1.556   0.75    60.00  70.00   
117788        306.0  72.22  1188000.0  1.514  10.80    83.33  66.50   
117789        289.4  72.22    32400.0  1.667   0.36     0.00  70.00   
117790        303.4  69.44    71229.0  1.583   0.79     0.00  69.25   
117791        301.3  81.94   312750.0  1.625   3.48    25.00  65.75   
117792        304.3  66.67  1620000.0  1.500  18.00    77.78  67.25   

statistic  Scrambling  year  month  
117776          63.33  2018      1  
117777          66.67  2018      2  
117778          70.83  2018      3  
117779          70.37  2018      3  
117780          70.37  2018      3  
117781          50.00  2018      4  
117782          71.43  2018      5  
117783          63.64  2018      5  
117784          66.67  2018      6  
117785          57.89  2018      7  
117786          66.67  2018      7  
117787          48.15  2018      8  
117788          70.00  2018      8  
117789          80.00  2018      8  
117790          72.73  2018      9  
117791          61.54  2018      9  
117792          70.83  2018      9

In [24]:
#How the 2018 winner performed in 2017
df_2018winner = df.loc[(df['player_name'] == 'Patrick Reed') & (df['year']==2017)]
df_2018winner.head(50)

statistic   player_name        date  \
94161      Patrick Reed  2017-01-08   
94162      Patrick Reed  2017-01-22   
94163      Patrick Reed  2017-02-05   
94164      Patrick Reed  2017-02-12   
94165      Patrick Reed  2017-02-19   
94166      Patrick Reed  2017-03-05   
94167      Patrick Reed  2017-03-12   
94169      Patrick Reed  2017-05-07   
94170      Patrick Reed  2017-05-14   
94171      Patrick Reed  2017-05-21   
94172      Patrick Reed  2017-06-04   
94173      Patrick Reed  2017-06-18   
94174      Patrick Reed  2017-06-25   
94175      Patrick Reed  2017-07-02   
94176      Patrick Reed  2017-07-09   
94177      Patrick Reed  2017-08-06   
94178      Patrick Reed  2017-08-13   
94179      Patrick Reed  2017-08-27   
94180      Patrick Reed  2017-09-04   
94181      Patrick Reed  2017-09-17   
94182      Patrick Reed  2017-09-24   
94183      Patrick Reed  2017-10-22   
94184      Patrick Reed  2017-10-22   

statistic                                         tournament  Fairways  \
94161                            SBS Tournament of Champions     81.67   
94162                                CareerBuilder Challenge     48.21   
94163                          Waste Management Phoenix Open     44.64   
94164                               AT&T Pebble Beach Pro-Am     76.36   
94165                                           Genesis Open     50.00   
94166           World Golf Championships-Mexico Championship     60.71   
94167                                   Valspar Championship     61.54   
94169                               Wells Fargo Championship     42.86   
94170                               THE PLAYERS Championship     67.86   
94171                                      AT&T Byron Nelson     42.86   
94172        the Memorial Tournament presented by Nationwide     66.07   
94173                                              U.S. Open     73.21   
94174                                 Travelers Championship     78.57   
94175                                 Quicken Loans National     48.21   
94176                                 The Greenbrier Classic     53.57   
94177      World Golf Championships-Bridgestone Invitational     35.71   
94178                                       PGA Championship     57.14   
94179                                     THE NORTHERN TRUST     69.64   
94180                         Dell Technologies Championship     55.36   
94181                                       BMW Championship     53.57   
94182                                      TOUR Championship     55.36   
94183                           Sanderson Farms Championship     55.36   
94184                              THE CJ CUP @ NINE BRIDGES     55.36   

statistic  Distance    GIR     Money  Putts  Purse  Sandies  Score  \
94161         279.3  83.33  210000.0  1.639   3.44    66.67  69.25   
94162         291.3  59.72  113680.0  1.417   1.96    60.00  68.75   
94163         308.1  61.11   13534.0  1.611   0.20    42.86  71.25   
94164         283.1  70.83   56520.0  1.625   0.79    42.86  70.25   
94165         282.5  55.56   15610.0  1.542   0.22    44.44  71.00   
94166         309.6  59.72   47250.0  1.611   0.48    28.57  72.50   
94167         291.6  59.72   28350.0  1.556   0.45    70.00  70.75   
94169         282.8  59.72  142500.0  1.528   1.90    33.33  70.75   
94170         290.1  68.06   92138.0  1.708   0.88    45.45  72.00   
94171         287.8  52.78   78214.0  1.389   1.04    77.78  69.00   
94172         293.8  62.50   19227.0  1.653   0.22    30.77  73.25   
94173         301.5  63.89  203557.0  1.556   1.70    50.00  70.50   
94174         312.9  73.61  248200.0  1.611   3.65    25.00  67.75   
94175         297.8  58.33   99400.0  1.583   1.40    50.00  69.50   
94176         309.1  59.72   69501.0  1.458   0.98    64.29  68.50   
94177         314.4  63.89   68000.0  1.569   0.70    25.00  70.25   
94178         302.6  59.72  784000.0  1.542   7.47    83.33  69.50   
94179         296.1  58.33

## Analyzing prior Masters Tournament top finishers

In [25]:
#2016
#df_2016_winners = df.loc[(df['tournament'] == 'Masters Tournament') & (df['year']==2016)]
#df_2016_winners.nsmallest(10, 'Score')

In [26]:
#2017
#Sergio Garcia won in the playoff
df_2017_winners = df.loc[(df['tournament'] == 'Masters Tournament') & (df['year']==2017)]
df_2017_winners.nsmallest(10, 'Score')

statistic       player_name        date          tournament  Fairways  \
70164           Justin Rose  2017-04-09  Masters Tournament     62.50   
110292        Sergio Garcia  2017-04-09  Masters Tournament     80.36   
21401      Charl Schwartzel  2017-04-09  Masters Tournament     69.64   
85495           Matt Kuchar  2017-04-09  Masters Tournament     67.86   
117410       Thomas Pieters  2017-04-09  Masters Tournament     60.71   
95037            Paul Casey  2017-04-09  Masters Tournament     62.50   
73210        Kevin Chappell  2017-04-09  Masters Tournament     60.71   
104543         Rory McIlroy  2017-04-09  Masters Tournament     51.79   
744              Adam Scott  2017-04-09  Masters Tournament     64.29   
106010           Ryan Moore  2017-04-09  Masters Tournament     75.00   

statistic  Distance    GIR      Money  Putts  Purse  Sandies  Score  \
70164         279.8  75.00  1188000.0  1.667  11.88     0.00  69.75   
110292        291.9  75.00  1980000.0  1.653  19.80    83.33  69.75   
21401         280.8  68.06   748000.0  1.625   7.48    60.00  70.50   
85495         269.9  62.50   484000.0  1.556   4.84    33.33  70.75   
117410        293.0  66.67        NaN  1.611   4.84    44.44  70.75   
95037         277.9  77.78   396000.0  1.764   3.96     0.00  71.00   
73210         288.5  70.83   354750.0  1.708   3.55    20.00  71.25   
104543        288.0  61.11   354750.0  1.611   3.55    42.86  71.25   
744           291.0  73.61   308000.0  1.722   3.08    57.14  71.50   
106010        276.8  63.89   308000.0  1.653   3.08    50.00  71.50   

statistic  Scrambling  year  month  
70164           50.00  2017      4  
110292          66.67  2017      4  
21401           56.52  2017      4  
85495           70.37  2017      4  
117410          58.33  2017      4  
95037           56.25  2017      4  
73210           42.86  2017      4  
104543          75.00  2017      4  
744             47.37  2017      4  
106010          61.54  2017      4

In [27]:
#2018
df_2018_winners = df.loc[(df['tournament'] == 'Masters Tournament') & (df['year']==2018)]
df_2018_winners.nsmallest(10, 'Score')

statistic     player_name        date          tournament  Fairways  Distance  \
94192        Patrick Reed  2018-04-08  Masters Tournament     73.21     299.3   
100693      Rickie Fowler  2018-04-08  Masters Tournament     71.43     290.6   
69000       Jordan Spieth  2018-04-08  Masters Tournament     67.86     287.1   
68287            Jon Rahm  2018-04-08  Masters Tournament     73.21     295.4   
18650        Bubba Watson  2018-04-08  Masters Tournament     83.93     304.0   
19797       Cameron Smith  2018-04-08  Masters Tournament     58.93     293.3   
50235      Henrik Stenson  2018-04-08  Masters Tournament     76.79     284.9   
104555       Rory McIlroy  2018-04-08  Masters Tournament     62.50     302.6   
80513       Marc Leishman  2018-04-08  Masters Tournament     53.57     291.8   
38593      Dustin Johnson  2018-04-08  Masters Tournament     66.07     304.9   

statistic    GIR      Money  Putts  Purse  Sandies  Score  Scrambling  year  \
94192      66.67  1980000.0  1.444  18.00    50.00  68.25       62.50  2018   
100693     70.83  1188000.0  1.569  10.80    50.00  68.50       76.19  2018   
69000      72.22   748000.0  1.625   6.80    66.67  68.75       70.00  2018   
68287      68.06   528000.0  1.556   4.80    75.00  69.25       60.87  2018   
18650      77.78   386375.0  1.694   3.51    60.00  69.75       62.50  2018   
19797      68.06   386375.0  1.583   3.51    40.00  69.75       69.57  2018   
50235      70.83   386375.0  1.625   3.51    50.00  69.75       57.14  2018   
104555     59.72   386375.0  1.514   3.51    80.00  69.75       75.86  2018   
80513      65.28   319000.0  1.569   2.90    66.67  70.00       72.00  2018   
38593      68.06   286000.0  1.667   2.60    50.00  70.25       65.22  2018   

statistic  month  
94192          4  
100693         4  
69000          4  
68287          4  
18650          4  
19797          4  
50235          4  
104555         4  
80513          4  
38593          4

In [28]:
#2019
df_2019_winners = df.loc[(df['tournament'] == 'Masters Tournament') & (df['year']==2019)]
df_2019_winners.nsmallest(10, 'Score')

statistic         player_name        date          tournament  Fairways  \
117798            Tiger Woods  2019-04-14  Masters Tournament     62.50   
17525           Brooks Koepka  2019-04-14  Masters Tournament     69.64   
38616          Dustin Johnson  2019-04-14  Masters Tournament     60.71   
127350      Xander Schauffele  2019-04-14  Masters Tournament     62.50   
40810      Francesco Molinari  2019-04-14  Masters Tournament     73.21   
54817               Jason Day  2019-04-14  Masters Tournament     66.07   
123063             Tony Finau  2019-04-14  Masters Tournament     67.86   
126086           Webb Simpson  2019-04-14  Masters Tournament     83.93   
68306                Jon Rahm  2019-04-14  Masters Tournament     76.79   
94045         Patrick Cantlay  2019-04-14  Masters Tournament     64.29   

statistic  Distance    GIR      Money  Putts  Purse  Sandies  Score  \
117798        294.6  80.56  2070000.0  1.667    NaN    66.67  68.75   
17525         313.6  73.61   858667.0  1.639    NaN   100.00  69.00   
38616         308.0  70.83   858667.0  1.569    NaN   100.00  69.00   
127350        305.8  70.83   858667.0  1.597    NaN    57.14  69.00   
40810         294.8  65.28   403938.0  1.458    NaN    66.67  69.25   
54817         296.5  70.83   403938.0  1.556    NaN    33.33  69.25   
123063        316.3  66.67   403938.0  1.556    NaN    63.64  69.25   
126086        283.1  68.06   403938.0  1.556    NaN    50.00  69.25   
68306         308.4  70.83   310500.0  1.694    NaN    71.43  69.50   
94045         299.9  61.11   310500.0  1.472    NaN    71.43  69.50   

statistic  Scrambling  year  month  
117798          50.00  2019      4  
17525           68.42  2019      4  
38616           80.95  2019      4  
127350          57.14  2019      4  
40810           84.00  2019      4  
54817           71.43  2019      4  
123063          79.17  2019      4  
126086          78.26  2019      4  
68306           80.95  2019      4  
94045           67.86  2019      4

## Reducing the data down to the last 2 years prior to the 2019 Masters (2017-2018)

In [29]:
df_2019winner = df[(df.year > 2016) & (df.year < 2019)]
df_2019winner.head()

statistic     player_name        date  \
2          A.J. McInerney  2017-11-05   
3          A.J. McInerney  2018-06-10   
220        Aaron Baddeley  2017-01-08   
221        Aaron Baddeley  2017-01-22   
222        Aaron Baddeley  2017-03-19   

statistic                                         tournament  Fairways  \
2                       Shriners Hospitals for Children Open     50.00   
3                                     FedEx St. Jude Classic     50.00   
220                              SBS Tournament of Champions     76.67   
221                                  CareerBuilder Challenge     50.00   
222        Arnold Palmer Invitational presented by Master...     53.57   

statistic  Distance    GIR    Money  Putts  Purse  Sandies  Score  Scrambling  \
2             317.8  76.39      NaN  1.694   2.22    22.22  69.50       52.94   
3             298.5  56.94      NaN  1.514   0.22    33.33  70.75       61.29   
220           287.9  81.94  70000.0  1.750   1.15    50.00  71.50       46.15   
221           283.6  55.56  10962.0  1.486   0.19    57.14  71.75       65.63   
222           285.9  58.33  33060.0  1.583   0.38    55.56  72.25       63.33   

statistic  year  month  
2          2017     11  
3          2018      6  
220        2017      1  
221        2017      1  
222        2017      3

## Create an Augusta National score

In [37]:
#TO DO AFTER: Create a score for players who perform better at Augusta National (where the Masters is played)
#TO DO AFTER: Look at stats of winners or top 5 Masters finishers and add that to the model
#TO DO AFTER: Look at career stats of players who have won the Masters before

In [38]:
#TO DO AFTER: Create a"trend" score by player. If they are improving / having a hot streak

## Creating a "Trending" score

## Aggregating the data by player

In [30]:
#TO DO AFTER: May only use the Total Sum of Money to the dataframe
#TO DO AFTER: And/or only use the Total Sum of Money won at prior Masters tournaments
aggregation_functions = {'Score': 'mean', 'Fairways': 'mean', 'Distance': 'mean', 'GIR': 'mean', 'Putts': 'mean', 'Sandies': 'mean', 'Scrambling': 'mean', 'Money': 'sum'} #'player_name': 'first'
df_2019winner_agg = df.groupby(df_2019winner['player_name']).aggregate(aggregation_functions)
df_2019winner_agg.nlargest(30, 'Money')
#df_2019winner_agg.head()

Score   Fairways    Distance        GIR     Putts  \
player_name                                                                 
Justin Thomas       68.925676  58.864324  311.280645  69.482162  1.575054   
Dustin Johnson      68.678571  58.024857  314.150000  70.754286  1.586171   
Brooks Koepka       69.570312  57.875000  311.696552  65.885625  1.581625   
Marc Leishman       69.672222  60.407556  299.272500  68.270444  1.605600   
Bryson DeChambeau   69.263889  63.123333  305.311111  69.598889  1.606361   
Justin Rose         69.340909  63.103939  301.712500  68.308182  1.591818   
Jordan Spieth       68.878571  60.152857  298.105714  70.595714  1.587686   
Rickie Fowler       69.283784  65.844054  300.402703  68.393514  1.585568   
Xander Schauffele   69.945946  60.922703  308.430000  67.228649  1.612162   
Tony Finau          69.500000  57.642041  312.706522  69.189184  1.605367   
Jason Day           69.784722  56.943611  307.800000  65.780000  1.566417   
Webb Simpson        69.226190  65.202619  290.788095  66.831667  1.572071   
Jon Rahm            69.209677  59.611935  308.323333  69.757097  1.599548   
Rory McIlroy        69.528846  54.685000  318.856000  65.918462  1.565115   
Patrick Reed        70.054878  57.508049  297.415789  63.279512  1.559976   
Hideki Matsuyama    69.955882  60.977059  301.640000  67.688529  1.607147   
Francesco Molinari  69.803030  67.872121  296.890625  68.896364  1.617030   
Gary Woodland       69.666667  62.704222  310.966667  71.728000  1.653978   
Brian Harman        70.090426  67.821489  290.333333  66.695957  1.596574   
Paul Casey          69.900000  65.037500  300.030303  69.027750  1.614600   
Charley Hoffman     70.166667  58.864375  299.308889  67.765833  1.614604   
Kyle Stanley        70.016667  70.927333  296.000000  70.987556  1.645333   
Kevin Kisner        69.711538  68.352821  289.910256  67.306923  1.589436   
Patrick Cantlay     69.705882  61.773235  303.118182  70.791471  1.636412   
Pat Perez           70.262500  65.031250  290.639394  67.534750  1.610400   
Keegan Bradley      70.053191  68.815106  298.833333  68.912553  1.626574   
Phil Mickelson      69.980263  53.776579  298.560526  64.765526  1.566579   
Adam Hadwin         70.250000  66.380000  290.510811  67.732326  1.600744   
Tiger Woods         69.264706  59.424706  304.082353  68.463529  1.576059   
Charles Howell III  69.972222  60.711556  299.051282  69.135111  1.630244   

                      Sandies  Scrambling       Money  
player_name                                            
Justin Thomas       55.189189   63.066757  18319086.0  
Dustin Johnson      50.920857   64.209143  16103128.0  
Brooks Koepka       49.109063   61.052187  14132752.0  
Marc Leishman       49.296667   61.625111  13339364.0  
Bryson DeChambeau   54.578611   61.037500  12871237.0  
Justin Rose         61.073030   63.352121  12039631.0  
Jordan Spieth       54.055143   62.823429  11761628.0  
Rickie Fowler       62.194324   64.517838  10710484.0  
Xander Schauffele   55.596486   60.195405   9697539.0  
Tony Finau          48.585510   60.791224   9452503.0  
Jason Day           53.787222   61.926667   9040289.0  
Webb Simpson        64.823810   67.015952   8871363.0  
Jon Rahm            55.318065   63.914839   8808619.0  
Rory McIlroy        61.405385   63.630769   8364093.0  
Patrick Reed        54.758537   63.207561   8226134.0  
Hideki Matsuyama    48.992647   63.072353   8033589.0  
Francesco Molinari  51.449697   62.431212   8004412.0  
Gary Woodland       45.740667   55.418889   7958208.0  
Brian Harman        58.362553   62.837021   7384256.0  
Paul Casey          43.803000   61.620750   7289017.0  
Charley Hoffman     53.388333   58.213333   7059798.0  
Kyle Stanley        48.904889   59.398667   7011419.0  
Kevin Kisner        51.020513   65.442564   6979833.0  
Patrick Cantlay     50.981765   59.815882   6703156.0  
Pat Perez           50.796000   62.439500   6503839.0  
Keegan Bradley      51.248298   59.611915   621

## Selecting independent variables

### Remove variables with high P-values

#### Comparing the variables to "Score" (average number of strokes per completed round)
None have P-values above 0.05

In [31]:
#Score_model = sm.OLS.from_formula("Score ~ Fairways + Distance + GIR + Putts + Sandies + Scrambling", data=df)
#Score_results = Score_model.fit()
#print("P-values:"), print(Score_results.pvalues), print("\nR-squared:"), print(Score_results.rsquared)
#print(Score_results.summary())

#Note:
#P-value is the probability of getting a sample proportion at least this extreme.  Ranges from zero to 100%. 
#NThe P>|t| value in the OLS summary reflects a two-sided P-value, and needs to be divided by 2 to reflect the single sided value.
#P-values quoted in the summary of this notebook reflect the single sided P-value

In [32]:
#Fairways
Score_Fairways_model = sm.OLS.from_formula("Score ~ Fairways", data=df_2019winner)
Score_Fairways_results = Score_Fairways_model.fit()
print("Score + Fairways | P-value:", Score_Fairways_results.pvalues[1]/2, "| R-squared:", Score_Fairways_results.rsquared)

#Distance
Score_Distance_model = sm.OLS.from_formula("Score ~ Distance", data=df_2019winner)
Score_Distance_results = Score_Distance_model.fit()
print("Score + Distance | P-value:", Score_Distance_results.pvalues[1]/2, "| R-squared:", Score_Distance_results.rsquared)

#GIR
Score_GIR_model = sm.OLS.from_formula("Score ~ GIR", data=df_2019winner)
Score_GIR_results = Score_GIR_model.fit()
print("Score + GIR | P-value:", Score_GIR_results.pvalues[1]/2, "| R-squared:", Score_GIR_results.rsquared)

#Putts
Score_Putts_model = sm.OLS.from_formula("Score ~ Putts", data=df_2019winner)
Score_Putts_results = Score_Putts_model.fit()
print("Score + Putts | P-value:", Score_Putts_results.pvalues[1]/2, "| R-squared:", Score_Putts_results.rsquared)

#Sandies
Score_Sandies_model = sm.OLS.from_formula("Score ~ Sandies", data=df_2019winner)
Score_Sandies_results = Score_Sandies_model.fit()
print("Score + Sandies | P-value:", Score_Sandies_results.pvalues[1]/2, "| R-squared:", Score_Sandies_results.rsquared)

#Scrambling
Score_Scrambling_model = sm.OLS.from_formula("Score ~ Scrambling", data=df_2019winner)
Score_Scrambling_results = Score_Scrambling_model.fit()
print("Score + Scrambling | P-value:", Score_Scrambling_results.pvalues[1]/2, "| R-squared:", Score_Scrambling_results.rsquared)

#Money
Score_Money_model = sm.OLS.from_formula("Score ~ Money", data=df_2019winner)
Score_Money_results = Score_Money_model.fit()
print("Score + Money | P-value:", Score_Money_results.pvalues[1]/2, "| R-squared:", Score_Money_results.rsquared)


#Note:
#P-value is the probability of getting a sample proportion at least this extreme.  Ranges from zero to 100%. 
#The P>|t| value in the OLS summary reflects a two-sided P-value, and needs to be divided by 2 to reflect the single sided value.
#P-values quoted in the summary of this notebook reflect the single sided P-value

Score + Fairways | P-value: 6.755827024546491e-24 | R-squared: 0.015276167455942735
Score + Distance | P-value: 1.4939572340541592e-53 | R-squared: 0.03875676287805219
Score + GIR | P-value: 0.0 | R-squared: 0.30789041017784424
Score + Putts | P-value: 7.804261609435294e-144 | R-squared: 0.09510223473481627
Score + Sandies | P-value: 3.610320314637166e-52 | R-squared: 0.03464898889070356
Score + Scrambling | P-value: 0.0 | R-squared: 0.20691257953965447
Score + Money | P-value: 6.654830057503988e-226 | R-squared: 0.15817635657748086


In [33]:
#Fairways
Money_Fairways_model = sm.OLS.from_formula("Money ~ Fairways", data=df_2019winner)
Money_Fairways_results = Money_Fairways_model.fit()
print("Money + Fairways | P-value:", Money_Fairways_results.pvalues[1]/2, "| R-squared:", Money_Fairways_results.rsquared)

#Distance
Money_Distance_model = sm.OLS.from_formula("Money ~ Distance", data=df_2019winner)
Money_Distance_results = Money_Distance_model.fit()
print("Money + Distance | P-value:", Money_Distance_results.pvalues[1]/2, "| R-squared:", Money_Distance_results.rsquared)

#GIR
Money_GIR_model = sm.OLS.from_formula("Money ~ GIR", data=df_2019winner)
Money_GIR_results = Money_GIR_model.fit()
print("Money + GIR | P-value:", Money_GIR_results.pvalues[1]/2, "| R-squared:", Money_GIR_results.rsquared)

#Putts
Money_Putts_model = sm.OLS.from_formula("Money ~ Putts", data=df_2019winner)
Money_Putts_results = Money_Putts_model.fit()
print("Money + Putts | P-value:", Money_Putts_results.pvalues[1]/2, "| R-squared:", Money_Putts_results.rsquared)

#Sandies
Money_Sandies_model = sm.OLS.from_formula("Money ~ Sandies", data=df_2019winner)
Money_Sandies_results = Money_Sandies_model.fit()
print("Money + Sandies | P-value:", Money_Sandies_results.pvalues[1]/2, "| R-squared:", Money_Sandies_results.rsquared)

#Scrambling
Money_Scrambling_model = sm.OLS.from_formula("Money ~ Scrambling", data=df_2019winner)
Money_Scrambling_results = Money_Scrambling_model.fit()
print("Money + Scrambling | P-value:", Money_Scrambling_results.pvalues[1]/2, "| R-squared:", Money_Scrambling_results.rsquared)

#Money
Money_Score_model = sm.OLS.from_formula("Money ~ Score", data=df_2019winner)
Money_Score_results = Money_Score_model.fit()
print("Money + Score | P-value:", Money_Score_results.pvalues[1]/2, "| R-squared:", Money_Score_results.rsquared)


Money + Fairways | P-value: 7.280336887542712e-12 | R-squared: 0.007605923676300974
Money + Distance | P-value: 4.1450150290418777e-26 | R-squared: 0.019741362357146075
Money + GIR | P-value: 4.201705289554516e-60 | R-squared: 0.04356901279311487
Money + Putts | P-value: 3.525722583745276e-49 | R-squared: 0.03551109224753557
Money + Sandies | P-value: 3.048092868881996e-18 | R-squared: 0.012396686769008736
Money + Scrambling | P-value: 7.830899713908358e-67 | R-squared: 0.04850247889759607
Money + Score | P-value: 6.654830057457795e-226 | R-squared: 0.15817635657748064


### Comparing correlation between variables

In [34]:
df_2019winner.corr()

statistic   Fairways  Distance       GIR     Money     Putts     Purse  \
statistic                                                                
Fairways    1.000000 -0.296388  0.395611  0.087212  0.245720  0.080446   
Distance   -0.296388  1.000000  0.136719  0.140504  0.095575  0.124414   
GIR         0.395611  0.136719  1.000000  0.208732  0.470330  0.237703   
Money       0.087212  0.140504  0.208732  1.000000 -0.188444  0.972194   
Putts       0.245720  0.095575  0.470330 -0.188444  1.000000 -0.214305   
Purse       0.080446  0.124414  0.237703  0.972194 -0.214305  1.000000   
Sandies    -0.023434 -0.033921 -0.011629  0.111340 -0.211138  0.125554   
Score      -0.123597 -0.196867 -0.554879 -0.397714  0.308387 -0.440232   
Scrambling  0.018306 -0.064917  0.034352  0.220233 -0.421329  0.255329   
year        0.037669  0.111118  0.113423  0.007273 -0.011549 -0.003466   
month       0.127673  0.160783  0.209912  0.020111  0.179669 -0.011263   

statistic    Sandies     Score  Scrambling      year     month  
statistic                                                       
Fairways   -0.023434 -0.123597    0.018306  0.037669  0.127673  
Distance   -0.033921 -0.196867   -0.064917  0.111118  0.160783  
GIR        -0.011629 -0.554879    0.034352  0.113423  0.209912  
Money       0.111340 -0.397714    0.220233  0.007273  0.020111  
Putts      -0.211138  0.308387   -0.421329 -0.011549  0.179669  
Purse       0.125554 -0.440232    0.255329 -0.003466 -0.011263  
Sandies     1.000000 -0.186142    0.312846  0.014900  0.007713  
Score      -0.186142  1.000000   -0.454876 -0.134290 -0.075664  
Scrambling  0.312846 -0.454876    1.000000  0.011306 -0.130450  
year        0.014900 -0.134290    0.011306  1.000000  0.007192  
month       0.007713 -0.075664   -0.130450  0.007192  1.000000

## Creating the model

In [35]:
#TO DO AFTER: Is Money a better predictor, or Score?  Or something else?  Add in some of the ideas above
Model = sm.OLS.from_formula("Money ~ GIR + Putts + Scrambling", data=df_2019winner)
Results = Model.fit()
print(Results.summary())

                            OLS Regression Results                            
Dep. Variable:                  Money   R-squared:                       0.162
Model:                            OLS   Adj. R-squared:                  0.162
Method:                 Least Squares   F-statistic:                     385.8
Date:                Sun, 06 Sep 2020   Prob (F-statistic):          4.66e-229
Time:                        08:14:09   Log-Likelihood:                -81139.
No. Observations:                5974   AIC:                         1.623e+05
Df Residuals:                    5970   BIC:                         1.623e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   8.133e+05   6.45e+04     12.604      0.0

In [36]:
df_2019winner['Prediction'] = 820300 + (9656.4244*df_2019winner.GIR) + (-.00009014*df_2019winner.Putts) + (1398.0181*df_2019winner.Scrambling)
df_2019winner.nlargest(10, 'Prediction') #use .nsmallest for lowest score

/Users/nicole/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


statistic         player_name        date              tournament  Fairways  \
73685                Kevin Na  2018-10-14            CIMB Classic     75.00   
304                Aaron Wise  2018-05-20       AT&T Byron Nelson     89.29   
71354          Keith Mitchell  2018-05-20       AT&T Byron Nelson     85.71   
762                Adam Scott  2018-05-20       AT&T Byron Nelson     91.07   
40796      Francesco Molinari  2018-07-01  Quicken Loans National     75.00   
105460         Russell Henley  2018-06-24  Travelers Championship     83.93   
1490         Anders Albertson  2018-11-18         The RSM Classic     76.79   
52911              J.J. Spaun  2017-08-20    Wyndham Championship     76.79   
40802      Francesco Molinari  2018-09-09        BMW Championship     73.21   
73675                Kevin Na  2018-05-20       AT&T Byron Nelson     82.14   

statistic  Distance    GIR      Money  Putts  Purse  Sandies  Score  \
73685           NaN  88.89    89320.0  1.667   1.28    71.43  67.75   
304           297.3  91.67  1386000.0  1.667  18.00     0.00  65.25   
71354         308.1  90.28   400400.0  1.722   5.20    50.00  66.25   
762           300.5  88.89   207900.0  1.722   2.70    66.67  67.25   
40796         304.3  86.11  1278000.0  1.597  18.00    83.33  64.75   
105460        290.8  86.11   234500.0  1.694   3.35   100.00  66.75   
1490          289.5  86.11    92960.0  1.736   1.45    75.00  67.25   
52911         307.3  88.89    78735.0  1.764   1.36    33.33  67.25   
40802         293.0  86.11   252000.0  1.653   2.80    33.33  66.00   
73675         284.9  86.11   257950.0  1.653   3.35    50.00  67.00   

statistic  Scrambling  year  month    Prediction  
73685           87.50  2018     10  1.800986e+06  
304             66.67  2018      5  1.798710e+06  
71354           71.43  2018      5  1.791942e+06  
762             75.00  2018      5  1.783511e+06  
40796           90.00  2018      7  1.777636e+06  
105460          90.00  2018      6  1.777636e+06  
1490            90.00  2018     11  1.777636e+06  
52911           62.50  2017      8  1.766036e+06  
40802           80.00  2018      9  1.763656e+06  
73675           80.00  2018      5  1.763656e+06